### PR TITLE
Feature/add rotation to draw ellipsoid

### LINKED ISSUE #7239: Add rotation parameter to skimage.draw.ellipsoid
### Description:

Hi!

I think it would be beneficial if you could add rotation parameter to the ellipsoid function, just like in case of the ellipse function.

### PR DESCRIPTION
**Pull request by pamitabh**
## Description

Implements rotation to draw.ellipsoid (closes #7239).

## Checklist

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpydoc/blob/main/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

For maintainers and optionally contributors, please refer to the [instructions](https://scikit-image.org/docs/stable/development/contribute.html#documenting-changes) on how to document this PR for the release notes.

```release-note
...
```

### CODE REVIEW

**Comment by pamitabh** [comment]
Hi, I would appreciate if any dev can give me pointers on why the label-check is failing even though I have added type labels to all functions in the updated script?

**Comment by jni** [comment]
Oh I'm so sorry for the confusion @pamitabh, that's not what that's about, that is for the core devs to add a label to the PR so that it can be included in the correct section in the release notes... :pray:

**Comment by jni** [comment]
I'm not sure why the mac and optional deps tests are failing...

```
FAILED filters/tests/test_thresholding.py::test_thresholds_dask_compatibility[threshold_triangle-41-43] - ValueError: Array chunk size or shape is unknown. 
```

@lagru @jarrodmillman any ideas? I notice there's a [new dask release](https://pypi.org/project/dask/#history) on PyPI, and earlier builds were passing, so probably unrelated to this PR...

**Comment by lagru** [comment]
Me neither. I've opened a tracking issue in https://github.com/scikit-image/scikit-image/issues/7491 and we merged a temporary fix (https://github.com/scikit-image/scikit-image/pull/7493) to get our CI working again. 

Also updated this PR branch to include the fix. @pamitabh, I recommend pulling this remote PR branch to your local branch before committing any local changes. :)

File: skimage/draw/draw3d.py [hunk 1]
```diff
@@ -30,15 +40,56 @@ def ellipsoid(a, b, c, spacing=(1.0, 1.0, 1.0), levelset=False):
...
+
     """
     if (a <= 0) or (b <= 0) or (c <= 0):
         raise ValueError('Parameters a, b, and c must all be > 0')
 
     offset = np.r_[1, 1, 1] * np.r_[spacing]
-
+    phi, theta, psi = rotation
+    # Define the standard rotation matrix for the given Euler angles defined for the coordinate system (x,y,z)
+    standard_rotation_matrix = np.array(
```
Comment (lagru) [inline skimage/draw/draw3d.py:68]: There's some duplication with calls to `np.cos(psi)`, `np.cos(phi)`, ... going on here. Can we use sensibly named intermediate variables here? Might be good to streamline the whole creation into its own little helper function to make it reusable and easier to reason about. What do you think?

**Review by lagru — COMMENTED** [review]
This new feature will also need one or two tests covering it. Could you add those to [test_draw3d.py](https://github.com/scikit-image/scikit-image/blob/0e97db25f3638b923be9b470fc5232403b4ffa95/skimage/draw/tests/test_draw3d.py)? :pray: You can have a look around in that file on how to do that.

File: skimage/draw/draw3d.py (hunk 1 above)
Comment (pamitabh) [reply skimage/draw/draw3d.py]: moved it to an auxiallary function. How does that look?

**Review by pamitabh — COMMENTED** [review]
> This new feature will also need one or two tests covering it. Could you add those to [test_draw3d.py](https://github.com/scikit-image/scikit-image/blob/0e97db25f3638b923be9b470fc5232403b4ffa95/skimage/draw/tests/test_draw3d.py)? :pray: You can have a look around in that file on how to do that. 

done! let me know if there's anything else that I should do.

File: skimage/draw/draw3d.py [hunk 2]
```diff
@@ -2,22 +2,81 @@
 from scipy.special import elliprg
 
 
-def ellipsoid(a, b, c, spacing=(1.0, 1.0, 1.0), levelset=False):
-    """Generate ellipsoid for given semi-axis lengths.
+def _create_rotation_matrix(rotation: tuple[float, float, float]) -> np.ndarray:
```
Comment (michaelbratsch) [inline skimage/draw/draw3d.py:5]: Why not use the scipy https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.from_euler.html to create the rotation matrix?

**Comment by pamitabh** [comment]
I tried using scipy's mentioned function but the sequence of rotation axes wasn't clear to me. I was mainly confused by the inability to define the standard z-x'-z rotation order commonly followed in euler rotation and its limited documentation didn't help. In the end, I prioritized clarity and better documentation, and just implemented the standard formula (see source) as my own matrix.

 Is there an advantage to using the scipy function vs defining our own?

________________________________

**Comment by michaelbratsch** [comment]
> I tried using scipy's mentioned function but the sequence of rotation axes wasn't clear to me. I was mainly confused by the inability to define the standard z-x'-z rotation order commonly followed in euler rotation and its limited documentation didn't help. In the end, I prioritized clarity and better documentation, and just implemented the standard formula (see source) as my own matrix. Is there an advantage to using the scipy function vs defining our own?

I did a test and got for
`_create_rotation_matrix(rotation=(alpha, beta, gamma))` and `Rotation.from_euler('zxz', angles=(-alpha, -beta, -gamma), degrees=False).as_matrix()` the same results. Generally, I think it is better to have less code which needs to be maintained. Scipy is anyway being used heavily in scikit-image so I would not reimplement something that is already present.

Anyway, thanks for implementing this feature.

**Comment by pamitabh** [comment]
> I did a test and got for
_create_rotation_matrix(rotation=(alpha, beta, gamma)) and Rotation.from_euler('zxz', angles=(-alpha, -beta, -gamma), degrees=False).as_matrix() the same results.

Sure some combination of arguments will give same results, but it's not conceptually clear. How/Why does it make sense that a rotation of (alpha, beta, gamma) from first principles (explicitly defining rotation matrix) is equivalent to scipy's rotation with negative of those angles? scipy's documentation doesn't clarify this either.

**Comment by michaelbratsch** [comment]
> Sure some combination of arguments will give same results, but it's not conceptually clear. How/Why does it make sense that a rotation of (alpha, beta, gamma) from first principles (explicitly defining rotation matrix) is equivalent to scipy's rotation with negative of those angles? scipy's documentation doesn't clarify this either.

The scipy documentation has a link to wikipedia. If you look at https://en.wikipedia.org/wiki/Euler_angles#Rotation_matrix you will find the case `Z_alphaX_betaZ_gamma` for Proper Euler angles. This is what you implemented in `_create_rotation_matrix` but in your code the some signs are switched. If you do it the way it is written in wikipedia then you do not need the negative signs and it is the same as the scipy implementation.

**Comment by pamitabh** [comment]
wikipedia does not count as a citation, and the book cited there is not available online so I couldn't verify the formula. I did not implement the formula mentioned in the 'proper euler' section of the wikipedia, my source for the formula (also commonly found in standard physics classical mechanics texts from Goldstein, Marion Thorton etc.):
> [1] Eq.2 and 6-14 in Weisstein, Eric W. "Euler Angles." From MathWorld--A Wolfram Web Resource.
>             https://mathworld.wolfram.com/EulerAngles.html 

The formula that I have used differs in more than the signs (see last row's order of sin and cos).

**Comment by michaelbratsch** [comment]
It is a matter of notational convention and there is not really a right and wrong. Definitely, the current implementation will work. I just have to point out, that the definition of the angles used here has a different sign than the definition of the angles used in scipy. Since the rest of the code in scikit-image relies heavily on scipy, I am not sure it is good to break this.

> wikipedia does not count as a citation

It is a quite common reference for scikit-image. I would not want to bet on how long [mathworks.com](https://mathworld.wolfram.com) is going to be online in the future.

**Comment by pamitabh** [comment]
> It is a matter of notational convention and there is not really a right and wrong. Definitely, the current implementation will work. 

I agree that there is no right or wrong in this, just better conceptual clarity. flipping signs of angles doesn't make sense to me. 

> I just have to point out, that the definition of the angles used here has a different sign than the definition of the angles used in scipy. Since the rest of the code in scikit-image relies heavily on scipy, I am not sure it is good to break this.

3D coordinate ref systems (crs) need to be well defined for things like rotations to make sense. Before I started work on this, I tried searching if such a crs already exists in skimage or napari world (#7239), infamously napari is left handed system! scipy never explicitly defines its coordinate system used for rotations (at least I couldn't find any!). I feel this as a good enough reason to not use scipy rotation. Additionally, Please let me know if this conflicts with any other crs defined or used in skimage.  

> It is a quite common reference for scikit-image. I would not want to bet on how long [mathworks.com](https://mathworld.wolfram.com/) is going to be available.

I have disagreed with this practice before. No one in the science community cites a tertiary source like wikipedia, hopefully that can change for scientific software. 

Wikipedia might be online longer but its content can change. As I said, the formula that I have used is present in most standard physics texts which can be easily found online, I can add those as references if needed. Ref: Eq. (4.47),  H. Goldstein, C. Poole and J. Safko, “Classical Mechanics,” 3rd Edition, Addison Wesley, Boston, 2002.

**Comment by michaelbratsch** [comment]
Looking at the `EuclideanTransform` from scikit-image, I realised that there is already an implementation to generate rotation matrices from Euler angles (always assumed it to be only 2D :astonished:), https://github.com/scikit-image/scikit-image/blob/fe57b47ba46e9e11c626b97a8a24cd4d2c41fba1/skimage/transform/_geometric.py#L1286-L1291

It comes with it's own implementation to generate the rotation matrix https://github.com/scikit-image/scikit-image/blob/fe57b47ba46e9e11c626b97a8a24cd4d2c41fba1/skimage/transform/_geometric.py#L1228
but uses only the default `axes` `0, 1, 2`.

The generated rotation matrix `EuclideanTransform(rotation=(a, b, c), dimensionality=3).rotation` is the same as you get with `Rotation.from_euler('XYZ', angles=(a, b, c), degrees=False).as_matrix()` (seems straight forward because of the default axis definition). Keep in mind that here intrinsic rotations are defined versus the extrinsic rotations as for example `_create_rotation_matrix`.

@pamitabh Would you agree that the drawing of the ellipsoid and the `EuclideanTransform` should use the same notation to define the angles?

**Comment by pamitabh** [comment]
> Looking at the EuclideanTransform from scikit-image, I realised that there is already an implementation to generate rotation matrices from Euler angles (always assumed it to be only 2D 😲), 

Great!

> @pamitabh Would you agree that the drawing of the ellipsoid and the EuclideanTransform should use the same notation to define the angles?

Yes! We should also include the chosen conventions in our user docs (something similar to the docs on [datatypes](https://scikit-image.org/docs/stable/user_guide/data_types.html)).

I think scipy's implementation has signs switched as it defines the axes differently (see discussion [here](https://github.com/scipy/scipy/issues/10959)) they seem to be using the matplotlib's definition (x - horizontally right, y vertically up, z out of plane)

**Comment by stefanv** [comment]
I would very much appreciate an added section in the user guide that makes our axes choices explicit (as well as examines them carefully). We have the opportunity to change our definition, if truly necessary, with skimage2.

+1 to refactoring whatever we can in terms of the SciPy rotation primitives; those were added fairly recently.

**Comment by stefanv** [comment]
W.r.t. references in the docs. Yes, it's annoying that Wikipedia is not a permanent source. On the other hand, it is often the best quick introduction to a topic, and mathematical pages, typically being uncontentious, tend to be more stable. It is also widely accessible, unlike many textbooks.

In short: ideal references are:

1. Publicly accessible.
2. Vetted by an expert.
3. Versioned / snapshotted.
4. Stable (available for a long time).

And (1) is so important that we sometimes gamble on (3) and (4).

**Comment by pamitabh** [comment]
> I would very much appreciate an added section in the user guide that makes our axes choices explicit (as well as examines them carefully). We have the opportunity to change our definition, if truly necessary, with skimage2.
> 
> +1 to refactoring whatever we can in terms of the SciPy rotation primitives; those were added fairly recently.

Is this a good place to have a discussion or a new pr request should be opened for this?

**Comment by pamitabh** [comment]
> W.r.t. references in the docs. Yes, it's annoying that Wikipedia is not a permanent source. On the other hand, it is often the best quick introduction to a topic, and mathematical pages, typically being uncontentious, tend to be more stable. It is also widely accessible, unlike many textbooks.
> 
> In short: ideal references are:
> 
>     1. Publicly accessible.
> 
>     2. Vetted by an expert.
> 
>     3. Versioned / snapshotted.
> 
>     4. Stable (available for a long time).
> 
> 
> And (1) is so important that we sometimes gamble on (3) and (4).
> 
> /cc @scikit-image/core to witness
I agree with your overall argument, minor correction: it should be "1 is so important that we sometimes gamble on 2, 3, and 4", as wikipedia can't be considered "vetted by an expert". 

In general, I don't have any issues with wikipedia *unless* other readily available, more reliable sources exist. If a simple online search is able to give a better source (wolfram mathworld for example) then that should be preferred to the wikipedia. At the very least, the cited wikipedia formula should have a primary source listed on the wiki page.

File: skimage/draw/tests/test_draw3d.py [hunk 3]
```diff
@@ -114,6 +121,94 @@ def test_ellipsoid_levelset():
         ]
     )
 
+    assert_allclose(test, expected)
+    assert_allclose(test_rotation, expected)
+    assert_allclose(test_anisotropic, expected)
+
+
+def test_ellipsoid_rotation_bool():
+    test = ellipsoid(a=2, b=1, c=1, rotation=(np.pi / 2, np.pi / 2, np.pi / 2))
```
Comment (lagru) [inline skimage/draw/tests/test_draw3d.py:130]:
```suggestion
    test = ellipsoid(a=2, b=1, c=3, rotation=(np.pi / 2, np.pi / 2, np.pi / 2))
```
What do you think about using three distinct lengths here? To make the test more restrictive.

File: skimage/draw/tests/test_draw3d.py
Comment (lagru) [inline skimage/draw/tests/test_draw3d.py:1]: As of now I think the function is only tested with rotation angles of `np.pi / 2`. Do you think it's worth covering different angles?

File: skimage/draw/draw3d.py (hunk 2 above)
Comment (lagru) [reply skimage/draw/draw3d.py:5]: With regards to using SciPy's `Rotation.from_euler` or our own `EuclideanTransform`, I concur that we generally prefer reusing code even if the definitions differ a bit.

I feel like using `EuclideanTransform` — if possible — would be the way to go here. Though, I'm not very firm on the mathematics, here. I'm curious what you think would be the best way to address this?

Also, would it make sense to replace 

https://github.com/scikit-image/scikit-image/blob/fe57b47ba46e9e11c626b97a8a24cd4d2c41fba1/skimage/transform/_geometric.py#L1228

with SciPy's `Rotation.from_euler` in a separate PR?

**Review by lagru — COMMENTED** [review]
Thanks for adding to the test suite!

**Comment by michaelbratsch** [comment]
@lagru I agree that it is a good idea to use SciPy's `Rotation.from_euler` for
https://github.com/scikit-image/scikit-image/blob/fe57b47ba46e9e11c626b97a8a24cd4d2c41fba1/skimage/transform/_geometric.py#L1228
and opened a pull-request for it https://github.com/scikit-image/scikit-image/pull/7503.

File: skimage/draw/tests/test_draw3d.py (hunk 3 above)
Comment (pamitabh) [reply skimage/draw/tests/test_draw3d.py:130]: I think that the present tests are fine: easy to quickly do back of the envelope verification, the other test parameters have redundancy too

**Comment by pamitabh** [comment]
Closing this pull request as the implementation of rotation requires making a decision about the CRS, which is in one of the todo tasks for skimage. Once that is done, this should be pretty straightforward to implement using the `EuclideanTransform` module.

**Comment by lagru** [comment]
Nevertheless, thank you @pamitabh for the contribution and the discussions driving this forward! :blush:

**Pull request closed without merging**